c10/util/Deprecated.h

```diff
@@ -65,7 +65,13 @@
 // with [[deprecated]] though (although, if you are on an old version
 // of gcc which doesn't understand attributes, you'll get a -Wattributes
 // error that it is ignored
+#if defined(__GNUC__) && __GNUC__ < 5
```

Comment:
Wouldn't it be more robust to do the test:

```
#if defined(__has_cpp_attribute)
#if __has_cpp_attribute(deprecated)
... use it ...
```

This is what the code I quoted you when you asked about this earlier does.

Reply:
ah, I see, I'll try that.

Reply:
hm... I think that's actually incorrect because what we're trying to handle is:
```
// nvcc has a bug where it doesn't understand __attribute__((deprecated))
// declarations even when the host compiler supports it.
```
As far as I can tell GCC 4.8.5 *does* support the deprecated attribute through `__attribute__((deprecated))` (it builds without warnings on non-CUDA hosts that run gcc 4.8.5).  __has_cpp_attribute itself was added to gcc in version 5, so the code would work but it'd just be based on the fact that __has_cpp_attribute was added in the same version that [[deprecated]] started working with nvcc. Technically deprecated became part of the standard under the [[]]'s attributes syntax in c++14:
https://en.cppreference.com/w/cpp/language/attributes

Reply:
`__has_cpp_attribute` is for testing the `[[foo]]` syntax, not the `__atribute__((foo))` syntax. So I'd say the behavior you describe here is correct, and this code will work, and for the right reasons too.

Reply:
not sure what you mean. [[deprecated]] is only standard in c++14 (see link) and __has_cpp_attribute isn't defined by gcc < 5.

e.g., below, __attribute__((deprecated)) works fine. I'm just going by the comment in Deprecated.h that it doesn't work with nvcc. __has_cpp_attribute isn't available.

```
(base) [bvaughan@devgpu005.ash6 ~] cat main.cpp
#include "stdio.h"

__attribute__((deprecated)) void foo() {
  printf("in deprecated function\n");
}

int main() {
#if defined(__has_cpp_attribute)
  printf("has __has_cpp_attribute\n");
#else
  printf("doesn't have __has_cpp_attribute\n");
#endif

  foo();
}

(base) [bvaughan@devgpu005.ash6 ~] g++ -std=c++11 main.cpp
main.cpp: In function ‘int main()’:
main.cpp:14:3: warning: ‘void foo()’ is deprecated (declared at main.cpp:3) [-Wdeprecated-declarations]
   foo();
   ^
main.cpp:14:7: warning: ‘void foo()’ is deprecated (declared at main.cpp:3) [-Wdeprecated-declarations]
   foo();
       ^
(base) [bvaughan@devgpu005.ash6 ~] ./a.out
doesn't have __has_cpp_attribute
in deprecated function
(base) [bvaughan@devgpu005.ash6 ~]
```